satpy/utils.py

```diff
@@ -538,3 +539,86 @@ def get_chunk_pixel_size():
     else:
         array_size = CHUNK_SIZE ** 2
     return array_size
+
+
+def check_file_protocols(filenames, storage_options=None):
+    """Check filenames for transfer protocols, convert to FSFile objects if possible."""
+    if storage_options is None:
+        storage_options = {}
+    if isinstance(filenames, dict):
+        return _check_file_protocols_for_dicts(filenames, storage_options)
+    return _check_file_protocols(filenames, storage_options)
+
+
+def _check_file_protocols_for_dicts(filenames, storage_options):
+    res = {}
+    for reader, files in filenames.items():
+        opts = storage_options.get(reader, {})
+        res[reader] = _check_file_protocols(files, opts)
+    return res
+
+
+def _check_file_protocols(filenames, storage_options):
+    local_files, remote_files, fs_files = _sort_files_to_local_remote_and_fsfiles(filenames)
+
+    if remote_files:
+        return local_files + fs_files + _filenames_to_fsfile(remote_files, storage_options)
+
+    return local_files + fs_files
+
+
+def _sort_files_to_local_remote_and_fsfiles(filenames):
+    from urllib.parse import urlparse
+
+    from satpy.readers import FSFile
+
+    local_files = []
+    remote_files = []
+    fs_files = []
+    for f in filenames:
+        if isinstance(f, FSFile):
+            fs_files.append(f)
+        elif urlparse(f).scheme in ('', 'file') or "\\" in f:
+            local_files.append(f)
+        else:
+            remote_files.append(f)
+    return local_files, remote_files, fs_files
+
+
+def _filenames_to_fsfile(filenames, storage_options):
+    import fsspec
+
+    from satpy.readers import FSFile
+
+    if filenames:
+        fsspec_files = fsspec.open_files(filenames, **storage_options)
+        return [FSFile(f) for f in fsspec_files]
+    return []
+
+
+def _get_storage_options_from_reader_kwargs(reader_kwargs):
+    if reader_kwargs is None:
+        return None, None
+    storage_options = reader_kwargs.pop('storage_options', None)
```

Comment:
`pop` is changing the dict in place, right? so why return `reader_kwargs`?

Reply:
It is, but as discussed above it's more clear that the dict has been changed if a new version is returned and collected to a new variable.